typed/declaration/schema-schema.ipldsch

```diff
@@ -0,0 +1,358 @@
+##
+## ```
+## {
+##   "type": "string"
+## }
+## ```
+##
+## Other types have more content.  Consider this example of a map type:
+##
+## ```
+## {
+##   "type": "map",
+##   "keyType": "String",
+##   "valueType": "Int"
+## }
+## ```
+##
+type Type union {
+	| TypeBool "bool"
+	| TypeString "string"
+	| TypeBytes "bytes"
+	| TypeInt "int"
+	| TypeFloat "float"
+	| TypeMap "map"
+	| TypeList "list"
+	| TypeLink "link"
+	| TypeUnion "union"
+	| TypeStruct "struct"
+	| TypeEnum "enum"
+} representation inline "kind"
+
+## TypeKind enumerates all the major kinds of type.
+## Notice this enum's members are the same as the set of strings used as
+## discriminants in the Type union.
+##
+## TODO: not actually sure we'll need to declare this.  Only usage is
+## in the Type union representation details?
+type TypeKind enum {
+	| "bool"
+	| "string"
+	| "bytes"
+	| "int"
+	| "float"
+	| "map"
+	| "list"
+	| "link"
+	| "union"
+	| "struct"
+	| "enum"
+}
+
+## RepresentationKind is similar to TypeKind, but includes only those concepts
+## which exist at the IPLD *Data Model* level.
+##
+## In other words, structs, unions, and enumerations are not present:
+## those concepts are introduced in the IPLD Schema system, and when serialized,
+## all of them must be transformable to one of these representation kinds
+## (e.g. a "struct" TypeKind will usually be transformed to a "map"
+## RepresentationKind; "enum" TypeKind are always "string" RepresentationKind;
+## and so on.)
+##
+## RepresentationKind strings are sometimes used to to indicate part of the
+## definition in the details of Type; for example, they're used describing
+## some of the detailed behaviors of a "kinded"-style union type.
+type RepresentationKind enum {
+	| "bool"
+	| "string"
+	| "bytes"
+	| "int"
+	| "float"
+	| "map"
+	| "list"
+	| "link"
+}
+
+## TypeBool describes a simple boolean type.
+## It has no details.
+##
+type TypeBool struct {}
+
+## TypeString describes a simple string type.
+## It has no details.
+##
+type TypeString struct {}
+
+## TypeBytes describes a simple byte array type.
+## It has no details.
+##
+type TypeBytes struct {}
+
+## TypeInt describes a simple integer numeric type.
+## It has no details.
+##
+type TypeInt struct {}
+
+## TypeFloat describes a simple floating point numeric type.
+## It has no details.
+##
+type TypeFloat struct {}
+
+## TypeMap describes a key-value map.
+## The keys and values of the map have some specific type of their own.
+##
+type TypeMap struct {
+	keyType TypeName # additionally, the referenced type must be reprkind==string.
+	valueType TypeTerm
+	valueNullable optional Bool
+}
+
+## TypeList describes a list.
+## The values of the list have some specific type of their own.
+##
+type TypeList struct {
+	valueType TypeTerm
+	valueNullable optional Bool
+}
+
+## TypeLink describes a hash linking to another object (a CID).
+##
+## REVIEW: this currently has no details... but possibly it should have a
```

Comment:
I am not aware of the implications of adding a type hint but from work I have done in the past in with [FHIR](http://hl7.org/fhir/) which similarly can link/embedded many different types with others, a type hint would have been very beneficial.

Reply:
:+1: :+1: :+1:  I think we should definitely have a type hint here... I'm only leaving it blank this far for lack of a fully finished idea of how exactly that should work.

Is it as simple as adding a `targetType optional TypeName` property?  Or should it be `TypeTerm` (so it can link directly to anonymous map types)?  Are there any *more* properties that could be needed?

In the DSL, this could perhaps be like `type Foo struct { fieldName optional Link<TargetType> }` -- spending (hopefully the last of) our sigil complexity budget for those angle brackets.

Reply:
+1 to type hints on links. I think that’s super important.